File: LeetCode/Easy/0997_Find_the_Town_Judge/Solution.py
## 2022-01-02
class Solution:
    def findJudge(self, n: int, trust: List[List[int]]) -> int:
        trusted = set([i for i in range(1, n+1)])
        trusted_by = [0 for i in range(0, n+1)]
        
        for t in trust:
            if t[0] in trusted:
                trusted.remove(t[0])
            trusted_by[t[1]] += 1
        
        for person in trusted:
            if trusted_by[person] == n-1:
                return person
        
        return -1

File: LeetCode/Easy/0997_Find_the_Town_Judge/Solution.cpp
// 2023-01-22 (in and out-degree of graph)
class Solution {
public:
    int findJudge(int n, vector<vector<int>>& trust) {
        /*
            input: n people, and a list of trust relationships
            output: the town judge's label, or -1 if doesn't exist

            approach:
            - seems like we want to track 2 things:
                - how many people a person trusts
                - how many people a person is trusted by
            - track using 2 vectors
            - iterate thru the vectors, and find one where # people trusts = 0 and
              # people trusted by = n-1, and return the index

            analysis:
            - time: O(n)
            - space: O(n)
        */
        int m = trust.size();
        vector<int> trusts(n+1), trustedBy(n+1);
        for (int i=0; i<m; i++) {
            trusts[trust[i][0]]++;
            trustedBy[trust[i][1]]++;
        }
        for (int i=1; i<=n; i++) {
            if (trusts[i] == 0 && trustedBy[i] == n-1)
                return i;
        }
        return -1;
    }
};